### PR TITLE
8338389: [JFR] Long strings should be added to the string pool

### PR DESCRIPTION
**Pull request by jaokim**
When committing JFR events with string data longer than 128 characters, the string isn't added to the JFR string pool. In scenarios with many events containing large string values, this can lead to very large JFR recordings. This is mostly the case for custom events containing f.i. SQL strings or other data. In the case where the string data consist of mostly duplicate data, adding them to the string pool has been shown to reduce the recording size by a factor of 10.

This change will add all longer strings (>128 characters) to the string pool. Strings of length 16-128 are still added using the same pre-cache methodology as before. The raised StringPool.MAX_LIMIT is used by [`jdk.jfr.internal.event.EventWriter::putString`](https://github.com/openjdk/jdk/blob/master/src/jdk.jfr/share/classes/jdk/jfr/internal/event/EventWriter.java) to determine which strings to add to the pool.

## Testing
- tier1-tier3
- jdk_jfr
- Internal testing with high load, using custom JFR events containing long strings have been performed

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338389](https://bugs.openjdk.org/browse/JDK-8338389): [JFR] Long strings should be added to the string pool (**Bug** - P3)

### Reviewers
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20596/head:pull/20596` \
`$ git checkout pull/20596`

Update a local copy of the PR: \
`$ git checkout pull/20596` \
`$ git pull https://git.openjdk.org/jdk.git pull/20596/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20596`

View PR using the GUI difftool: \
`$ git pr show -t 20596`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20596.diff">https://git.openjdk.org/jdk/pull/20596.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20596#issuecomment-2291262962)

### CODE REVIEW

File: test/jdk/jdk/jfr/jvm/TestLongStringsInPool.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        // string. With long strings (>128 chars) being pooled,
+        // the two recording should be roughly the same size.
+        final int numEvents = 10;
+        final String longString = generateString();
+        final int strLen = longString.length();
+        final StringEvent event = new StringEvent();
+        event.message = longString;
+
+        Recording firstRec = new Recording();
+        firstRec.enable(StringEvent.class);
```
Comment (mgronlun) [inline test/jdk/jdk/jfr/jvm/TestLongStringsInPool.java:61]: No need to explicitly enable user-define events. They are turned on by default.

File: test/jdk/jdk/jfr/jvm/TestLongStringsInPool.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        firstRec.start();
+        // commit events with empty message (both recordings
+        // will have the same number of events
```
Comment (mgronlun) [inline test/jdk/jdk/jfr/jvm/TestLongStringsInPool.java:64]: Missing ")".

**Review by mgronlun — APPROVED** [review]

File: test/jdk/jdk/jfr/jvm/TestLongStringsInPool.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        event.commit();
+
+        secondRec.stop();
+        Path rec2 = Paths.get(".", "rec2.jfr");
+        secondRec.dump(rec2);
+        secondRec.close();
+
+        // the files aren't exactly the same size, but rec2 should
+        // not take up space for all strings if they're pooled correctly
+        long maxAllowedDiff = (numEvents - 1)  * strLen;
```
Comment (turbanoff) [inline test/jdk/jdk/jfr/jvm/TestLongStringsInPool.java:96]:
```suggestion
        long maxAllowedDiff = (numEvents - 1) * strLen;
```

**Comment by jaokim** [comment]
I've fixed the nits for the test, thanks @turbanoff and @mgronlun!

This now needs a re-review. Thanks!

**Review by mgronlun — APPROVED** [review]

**Comment by jaokim** [comment]
Thank you @mgronlun! 
Could I also request a sponsor for this? Thanks!

/integrate

**Pull request closed without merging**